### PR TITLE
KTOR-3178 Drop Redundant Checkcast from SFG Loop

### PR DESCRIPTION
**Pull request by e5l**
Fix [KTOR-3178](https://youtrack.jetbrains.com/issue/KTOR-3178)

### CODE REVIEW

File: ktor-utils/common/src/io/ktor/util/pipeline/Pipeline.kt [hunk 1]
```diff
@@ -45,10 +47,14 @@ public open class Pipeline<TSubject : Any, TContext : Any>(
     public val isEmpty: Boolean
         get() = interceptorsQuantity == 0
 
-    private val _interceptors: AtomicRef<List<suspend PipelineContext<TSubject, TContext>.(TSubject) -> Unit>?> =
+
+    private val _interceptorsArray: AtomicRef<Array<PipelineInterceptorFunction<TSubject, TContext>>?> =
```
Comment (rsinukov) [inline ktor-utils/common/src/io/ktor/util/pipeline/Pipeline.kt:51]: is it unused?

File: ktor-utils/common/src/io/ktor/util/pipeline/SuspendFunctionGun.kt [hunk 2]
```diff
@@ -14,50 +13,28 @@ import kotlin.coroutines.intrinsics.*
...
-                        // and null should be already returned at the beginning of the function.
-                        // So if we are here and see that the list is empty but lastPeekedIndex wasn't -1,
-                        // then it means that element(s) has disappeared concurrently.
-                        return StackWalkingFailedFrame
-                    }
-                    @Suppress("UNCHECKED_CAST")
-                    return peekContinuationFromList(rootContinuation as List<Continuation<*>>)
-                }
-                else -> return null
+            if (currentIndex == null) currentIndex = lastSuspensionIndex
```
Comment (rsinukov) [inline ktor-utils/common/src/io/ktor/util/pipeline/SuspendFunctionGun.kt:33]: is it OK to have auto boxing here?

**Review by rsinukov — CHANGES_REQUESTED** [review]

**Comment by rsinukov** [comment]
Please check compilation for native and js targets

File: ktor-utils/common/src/io/ktor/util/pipeline/Pipeline.kt (hunk 1 above)
Comment (e5l) [reply ktor-utils/common/src/io/ktor/util/pipeline/Pipeline.kt:51]: Sure

**Comment by rsinukov** [comment]
Builds are failing for natives with
```
startCoroutineUninterceptedOrReturn3
```

File: ktor-utils/common/src/io/ktor/util/pipeline/SuspendFunctionGun.kt [hunk 3]
```diff
@@ -19,20 +19,19 @@ internal class SuspendFunctionGun<TSubject : Any, TContext : Any>(
...
+        var currentIndex: Int = Int.MIN_VALUE
 
         override fun getStackTraceElement(): StackTraceElement? = null
 
         private fun peekContinuation(): Continuation<*>? {
-            if (currentIndex == null) currentIndex = lastSuspensionIndex
-            if (currentIndex!! < 0) {
-                currentIndex = null
+            if (currentIndex == Int.MIN_VALUE) currentIndex = lastSuspensionIndex
+            if (currentIndex < 0) {
```
Comment (rsinukov) [inline ktor-utils/common/src/io/ktor/util/pipeline/SuspendFunctionGun.kt:33]: When `currentIndex == Int.MIN_VALUE`, execution will go under this branch. Is it ok? If not, we can change the special case to `Int.MAX_VALUE`

**Review by rsinukov — APPROVED** [review]
LGTM, but please update native and js targets

File: ktor-utils/common/src/io/ktor/util/pipeline/SuspendFunctionGun.kt (hunk 3 above)
Comment (e5l) [reply ktor-utils/common/src/io/ktor/util/pipeline/SuspendFunctionGun.kt:33]: It's not possible to see `MIN_VALUE` here(see the previous line)

**Pull request merged**